File: packages/go/internal/seguranca/ataques/ataque_distribuicao_bytes.go
package ataques

import (
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesDistribuicaoBytes configura o AtaqueDistribuicaoBytes.
type OpcoesDistribuicaoBytes struct {
	AmostrasDeBlocos int
	TamanhoBloco     int
}

// AtaqueDistribuicaoBytes verifica se os bytes do keystream estão
// uniformemente distribuídos entre 0-255 (qui-quadrado).
type AtaqueDistribuicaoBytes struct {
	amostrasDeBlocos int
	tamanhoBloco     int
}

// NovoAtaqueDistribuicaoBytes cria o ataque. Sem argumento, usa
// amostrasDeBlocos = 500 e tamanhoBloco = 32.
func NovoAtaqueDistribuicaoBytes(opcoes ...OpcoesDistribuicaoBytes) *AtaqueDistribuicaoBytes {
	o := OpcoesDistribuicaoBytes{AmostrasDeBlocos: 500, TamanhoBloco: 32}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueDistribuicaoBytes{amostrasDeBlocos: o.AmostrasDeBlocos, tamanhoBloco: o.TamanhoBloco}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueDistribuicaoBytes) Nome() string {
	return "Distribuição de bytes (qui-quadrado)"
}

// Executar roda o ataque.
func (a *AtaqueDistribuicaoBytes) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	chave := []byte(alvo.ChaveDeTeste())

	contagem := make([]int, 256)
	total := 0
	for i := 0; i < a.amostrasDeBlocos; i++ {
		ks := alvo.GerarKeystreamBruto(chave, seguranca.RandomBytes(alvo.TamanhoIv()), "enc", a.tamanhoBloco)
		for j := 0; j < len(ks); j++ {
			contagem[ks[j]]++
			total++
		}
	}

	esperado := float64(total) / 256
	qui2 := 0.0
	for _, c := range contagem {
		diff := float64(c) - esperado
		qui2 += (diff * diff) / esperado
	}

	vulneravel := qui2 > 330
	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevMedia
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"qui-quadrado=%.1f sobre %d bytes (255 graus de liberdade; >330 é suspeito)",
			qui2, total,
		),
		Dados: map[string]any{"qui2": qui2},
	}, nil
}

File: packages/rust/src/seguranca/ataques/ataque_distribuicao_bytes.rs
//! Um keystream de qualidade deve ter bytes distribuídos uniformemente entre
//! 0-255. Um viés forte (qui-quadrado muito alto) indica que alguns valores
//! de byte saem com mais frequência que outros - sinal de fraqueza estatística
//! na função de mistura.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueDistribuicaoBytes {
    pub amostras_de_blocos: usize,
    pub tamanho_bloco: usize,
}

impl AtaqueDistribuicaoBytes {
    pub fn new() -> Self {
        AtaqueDistribuicaoBytes {
            amostras_de_blocos: 500,
            tamanho_bloco: 32,
        }
    }
}

impl Default for AtaqueDistribuicaoBytes {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueDistribuicaoBytes {
    fn nome(&self) -> String {
        "Distribuição de bytes (qui-quadrado)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let chave = alvo.chave_de_teste();

        let mut contagem = [0u64; 256];
        let mut total = 0u64;
        for _ in 0..self.amostras_de_blocos {
            let ks = alvo.gerar_keystream_bruto(
                chave.as_bytes(),
                &random_bytes(alvo.tamanho_iv()),
                "enc",
                self.tamanho_bloco,
            );
            for &b in &ks {
                contagem[b as usize] += 1;
                total += 1;
            }
        }

        let esperado = total as f64 / 256.0;
        let mut qui2 = 0.0f64;
        for &c in &contagem {
            qui2 += (c as f64 - esperado).powi(2) / esperado;
        }

        // Com 255 graus de liberdade, valores acima de ~330 já são
        // estatisticamente suspeitos (p < 0.01); acima de ~380, bem suspeitos.
        let vulneravel = qui2 > 330.0;

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Media
            } else {
                Severidade::Info
            },
            &format!(
                "qui-quadrado={:.1} sobre {} bytes (255 graus de liberdade; >330 é suspeito)",
                qui2, total
            ),
        ))
    }
}
